input/listen.go

```diff
@@ -1,40 +1,62 @@
 package input
 
-import "net"
+import (
+	"bytes"
+	"net"
+)
 
-func listen(addr string, handler Handler) (net.Listener, error) {
+func listen(addr string, handler Handler) error {
 	laddr, err := net.ResolveTCPAddr("tcp", addr)
 	if err != nil {
-		return nil, err
+		return err
 	}
 	l, err := net.ListenTCP("tcp", laddr)
 	if err != nil {
-		return nil, err
+		return err
 	}
 	log.Notice("listening on %v/tcp", laddr)
-	go accept(l, handler)
+	go acceptTcp(l, handler)
 
 	udp_addr, err := net.ResolveUDPAddr("udp", addr)
 	if err != nil {
-		return nil, err
+		return err
 	}
 	udp_conn, err := net.ListenUDP("udp", udp_addr)
 	if err != nil {
-		return nil, err
+		return err
 	}
 	log.Notice("listening on %v/udp", udp_addr)
-	go handler.Handle(udp_conn)
+	go acceptUdp(udp_conn, handler)
 
-	return l, nil
+	return nil
 }
 
-func accept(l *net.TCPListener, handler Handler) {
+func acceptTcp(l *net.TCPListener, handler Handler) {
 	for {
 		c, err := l.AcceptTCP()
 		if err != nil {
 			log.Error(err.Error())
 			break
 		}
-		go handler.Handle(c)
+		go acceptTcpConn(c, handler)
+	}
+}
+
+func acceptTcpConn(c net.Conn, handler Handler) {
+	defer c.Close()
+	log.Debug("listen.go: tcp connection from %v", c.RemoteAddr())
+	handler.Handle(c)
+}
+
+func acceptUdp(l *net.UDPConn, handler Handler) {
+	buffer := make([]byte, 65535)
+	for {
+		b, addr, err := l.ReadFrom(buffer)
+		if err != nil {
+			log.Error(err.Error())
```

Comment:
hmm so the moment we hit an error reading a single packet, we stop accepting any udp data. seems like we should handle this more gracefully?

Reply:
This hasn't changed, the existing code behaves in exactly the same way.

Reply:
To be clear, I do think that we should handle errors better here, since right now it's possiuble for crng to be running but no longer consuming via tcp and/or udp, but I don't want to expand this PR even further, so I'd like to deal with that separately.